File: sdk/python/_cortexdb_client/aql.py
from __future__ import annotations


def _quote_aql_string(value: str) -> str:
    escaped = (
        value.replace("\\", "\\\\")
        .replace('"', '\\"')
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
    )
    return f'"{escaped}"'


def _validate_aql_identifier(field: str, value: str) -> None:
    if not value:
        raise ValueError(f"{field} must be a non-empty AQL identifier")
    first = value[0]
    if not (first == "_" or first.isascii() and first.isalpha()):
        raise ValueError(f"{field} must start with '_' or an ASCII letter")
    for character in value[1:]:
        if not (
            character == "_"
            or character == "-"
            or character == ":"
            or character.isascii()
            and character.isalnum()
        ):
            raise ValueError(f"{field} contains an invalid AQL identifier character")


def _validate_decimal(field: str, value: str | None) -> None:
    if value is None:
        return
    left, separator, right = value.partition(".")
    if separator != "." or not left.isdigit() or not right.isdigit():
        raise ValueError(f"{field} must be a decimal literal")


def build_retrieve_context_aql(
    task: str,
    brain: str,
    *,
    mode: str | None = None,
    budget_tokens: int | None = None,
    limit_candidates: int | None = None,
    where_clause: str | None = None,
    require_citations: bool = False,
    min_confidence: str | None = None,
    source_trust: str | None = None,
    freshness_seconds: int | None = None,
    explain: bool = False,
) -> str:
    _validate_aql_identifier("brain", brain)
    if mode is not None and mode not in {"fast", "balanced", "semantic", "audit"}:
        raise ValueError("mode must be fast, balanced, semantic, or audit")
    if where_clause is not None and not where_clause.strip():
        raise ValueError("where_clause must not be empty")
    _validate_decimal("min_confidence", min_confidence)
    _validate_decimal("source_trust", source_trust)

    parts = []
    if explain:
        parts.append("EXPLAIN")
    parts.extend(["RETRIEVE CONTEXT FOR TASK", _quote_aql_string(task), "IN BRAIN", brain])
    if mode is not None:
        parts.extend(["USING MODE", mode])
    if budget_tokens is not None:
        parts.extend(["BUDGET", str(budget_tokens), "TOKENS"])
    if limit_candidates is not None:
        parts.extend(["LIMIT", str(limit_candidates), "CANDIDATES"])
    if where_clause is not None:
        parts.extend(["WHERE", where_clause.strip()])
    if require_citations:
        parts.extend(["REQUIRE", "citations"])
    if min_confidence is not None:
        parts.extend(["REQUIRE", "confidence", ">=", min_confidence])
    if source_trust is not None:
        parts.extend(["REQUIRE", "source_trust", ">=", source_trust])
    if freshness_seconds is not None:
        parts.extend(["REQUIRE", "freshness", "<=", str(freshness_seconds), "SECONDS"])
    return " ".join(parts) + ";"


def build_verify_fact_aql(fact: str, brain: str) -> str:
    _validate_aql_identifier("brain", brain)
    return f"VERIFY FACT {_quote_aql_string(fact)} IN BRAIN {brain};"


def build_remember_aql(
    content: str,
    scope: str,
    memory_type: str,
    *,
    ttl_seconds: int | None = None,
) -> str:
    _validate_aql_identifier("scope", scope)
    _validate_aql_identifier("memory_type", memory_type)
    statement = (
        f"REMEMBER {_quote_aql_string(content)} IN SCOPE {scope} AS TYPE {memory_type}"
    )
    if ttl_seconds is not None:
        statement += f" TTL {ttl_seconds} SECONDS"
    return statement + ";"

File: sdk/typescript/cortexdb-client/aql.ts
import type { AqlRetrievalMode, RetrieveContextAqlOptions } from "./types";

function quoteAqlString(value: string): string {
  return `"${value
    .replaceAll("\\", "\\\\")
    .replaceAll("\"", "\\\"")
    .replaceAll("\n", "\\n")
    .replaceAll("\r", "\\r")
    .replaceAll("\t", "\\t")}"`;
}

function validateAqlIdentifier(field: string, value: string): void {
  if (!/^[A-Za-z_][A-Za-z0-9_:-]*$/.test(value)) {
    throw new Error(`${field} must be an AQL identifier`);
  }
}

function validateDecimal(field: string, value: string | undefined): void {
  if (value !== undefined && !/^[0-9]+\.[0-9]+$/.test(value)) {
    throw new Error(`${field} must be a decimal literal`);
  }
}

function validateAqlMode(value: AqlRetrievalMode | undefined): void {
  if (
    value !== undefined &&
    value !== "fast" &&
    value !== "balanced" &&
    value !== "semantic" &&
    value !== "audit"
  ) {
    throw new Error("mode must be fast, balanced, semantic, or audit");
  }
}

export function buildRetrieveContextAql(
  task: string,
  brain: string,
  options: RetrieveContextAqlOptions = {},
): string {
  validateAqlIdentifier("brain", brain);
  validateAqlMode(options.mode);
  if (options.whereClause !== undefined && options.whereClause.trim().length === 0) {
    throw new Error("whereClause must not be empty");
  }
  validateDecimal("minConfidence", options.minConfidence);
  validateDecimal("sourceTrust", options.sourceTrust);

  const parts: string[] = [];
  if (options.explain) parts.push("EXPLAIN");
  parts.push("RETRIEVE CONTEXT FOR TASK", quoteAqlString(task), "IN BRAIN", brain);
  if (options.mode) parts.push("USING MODE", options.mode);
  if (options.budgetTokens !== undefined) {
    parts.push("BUDGET", String(options.budgetTokens), "TOKENS");
  }
  if (options.limitCandidates !== undefined) {
    parts.push("LIMIT", String(options.limitCandidates), "CANDIDATES");
  }
  if (options.whereClause !== undefined) {
    parts.push("WHERE", options.whereClause.trim());
  }
  if (options.requireCitations) parts.push("REQUIRE", "citations");
  if (options.minConfidence !== undefined) {
    parts.push("REQUIRE", "confidence", ">=", options.minConfidence);
  }
  if (options.sourceTrust !== undefined) {
    parts.push("REQUIRE", "source_trust", ">=", options.sourceTrust);
  }
  if (options.freshnessSeconds !== undefined) {
    parts.push("REQUIRE", "freshness", "<=", String(options.freshnessSeconds), "SECONDS");
  }
  return `${parts.join(" ")};`;
}

export function buildVerifyFactAql(fact: string, brain: string): string {
  validateAqlIdentifier("brain", brain);
  return `VERIFY FACT ${quoteAqlString(fact)} IN BRAIN ${brain};`;
}

export function buildRememberAql(
  content: string,
  scope: string,
  memoryType: string,
  ttlSeconds?: number,
): string {
  validateAqlIdentifier("scope", scope);
  validateAqlIdentifier("memoryType", memoryType);
  let statement = `REMEMBER ${quoteAqlString(content)} IN SCOPE ${scope} AS TYPE ${memoryType}`;
  if (ttlSeconds !== undefined) statement += ` TTL ${ttlSeconds} SECONDS`;
  return `${statement};`;
}

